### PR TITLE
Rename types in preparation for TypeDoc

### PR DESCRIPTION
**Pull request by lolmaus**
~⚠️ This PR contains commits from https://github.com/embroider-build/scenario-tester/pull/14 and #15. The diff will clear up once those are merged.~

This PR is in preparation for TypeDoc documentation.

This PR contains two changes:
* Renames the `root` callbacks and properties to `callbackCreateProject`. This makes them much more obvious, and also removes the ambiguity with `root` from `Project` where it's a string.
    It also defines a type for the callback, as it's used in many places throughout the code.
    And renames `ProjectMutator` type to `CallbackMutateProject` for consistency.
* Splits the `State` type into `ScenariosStateRoot` and `ScenariosStateDerived`. This lets me document them separately. The `Scenraios` prefix scopes the type to its relevant class.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -4,38 +4,69 @@ import { spawn } from 'child_process';
...
-type State =
-  | {
-      type: 'root';
-      root: () => Project | Promise<Project>;
-    }
-  | {
-      type: 'derived';
-      parent: Scenarios;
-      variants: Record<string, ProjectMutator[]>;
-    };
```
Comment (ef4) [inline index.ts:20]: As we discussed on the office hours, this type is entirely private and self-contained so it should go back to being not exported.

If we add documentation to it, the most useful documentation would be on its individual fields. The object itself is really just an help to typescript so we know which fields are available in which states.

Comment (lolmaus) [reply index.ts:20]: @ef4 Done!

File: index.ts [hunk 2]
```diff
@@ -4,38 +4,57 @@ import { spawn } from 'child_process';
...
-type ProjectMutator = (project: Project) => void | Promise<void>;
+/**
+ * A callback used by some of classes to obtain a {@link Project}.
+ * 
+ * You are supposed to pass such a callback. In it, you should create and return a {@link Project}
+ * instance.
+ * 
+ * @returns A {@link Project} instance, optionally wrapped into a promise.
+ */
+
```
Comment (mansona) [inline index.ts:15]:
```suggestion
```

shouldn't this comment block be "connected" to the following code for it to be picked up? or is it smart enough to know that it refers to the next line of code that isn't whitespace?

**Review by mansona — APPROVED** [review]
This looks good to me, it feels like the change of keys in some of the objects here are all entirely internal so it's not a breaking change.

I'm approving but I'm not 100% sure I should merge without someone who ~likes~ knows typescript having a quick look 👍

**Review by NullVoxPopuli — APPROVED** [review]
TS changes look good to me 💯

**Pull request merged**